exp/controllers/ocimachinepool_controller.go

```diff
@@ -0,0 +1,384 @@
+
+	// Find existing Instance Pool
+	instancePool, err := machinePoolScope.FindInstancePool(ctx)
+	if err != nil {
+		conditions.MarkUnknown(machinePoolScope.OCIMachinePool, infrav1exp.InstancePoolReadyCondition, infrav1exp.InstancePoolNotFoundReason, err.Error())
+		return ctrl.Result{}, err
+	}
+
+	if instancePool == nil {
+		if _, err := machinePoolScope.CreateInstancePool(ctx); err != nil {
+			conditions.MarkFalse(machinePoolScope.OCIMachinePool, infrav1exp.InstancePoolReadyCondition, infrav1exp.InstancePoolProvisionFailedReason, clusterv1.ConditionSeverityError, err.Error())
+			return ctrl.Result{}, err
+		}
+		r.Recorder.Eventf(machinePoolScope.OCIMachinePool, "SuccessfulCreate", "Created new Instance Pool: %s", machinePoolScope.OCIMachinePool.GetName())
+		return ctrl.Result{}, nil
+	}
+
+	if err := machinePoolScope.UpdatePool(ctx, instancePool); err != nil {
+		r.Recorder.Eventf(machinePoolScope.OCIMachinePool, corev1.EventTypeWarning, "FailedUpdate", "Failed to update instance pool: %v", err)
+		machinePoolScope.Error(err, "error updating OCIMachinePool")
+		return ctrl.Result{}, err
+	}
+
+	machinePoolScope.Info("OCI Compute Instance Pool found", "InstancePoolId", *instancePool.Id)
+	machinePoolScope.OCIMachinePool.Spec.ProviderID = common.String(fmt.Sprintf("oci://%s", *instancePool.Id))
+
+	switch instancePool.LifecycleState {
+	case core.InstancePoolLifecycleStateProvisioning, core.InstancePoolLifecycleStateStarting:
+		machinePoolScope.Info("Instance Pool is pending")
+		conditions.MarkFalse(machinePoolScope.OCIMachinePool, infrav1exp.InstancePoolReadyCondition, infrav1exp.InstancePoolNotReadyReason, clusterv1.ConditionSeverityInfo, "")
+		return reconcile.Result{RequeueAfter: 10 * time.Second}, nil
+	case core.InstancePoolLifecycleStateRunning:
+		machinePoolScope.Info("Instance pool is active")
+
+		// record the event only when pool goes from not ready to ready state
+		r.Recorder.Eventf(machinePoolScope.OCIMachinePool, corev1.EventTypeNormal, "InstancePoolReady",
+			"Instance pool is in ready state")
+		conditions.MarkTrue(machinePoolScope.OCIMachinePool, infrav1exp.InstancePoolReadyCondition)
+		machinePoolScope.SetReady()
+	default:
+		conditions.MarkFalse(machinePoolScope.OCIMachinePool, infrav1exp.InstancePoolReadyCondition, infrav1exp.InstancePoolProvisionFailedReason, clusterv1.ConditionSeverityError, "")
+		machinePoolScope.SetFailureReason(capierrors.CreateMachineError)
+		machinePoolScope.SetFailureMessage(errors.Errorf("Instance Pool status %q is unexpected", instancePool.LifecycleState))
+		r.Recorder.Eventf(machinePoolScope.OCIMachinePool, corev1.EventTypeWarning, "ReconcileError",
+			"Instance pool has invalid lifecycle state %s", instancePool.LifecycleState)
+		return reconcile.Result{}, errors.New(fmt.Sprintf("instance pool has invalid lifecycle state %s", instancePool.LifecycleState))
+	}
+
+	return ctrl.Result{}, nil
```

Comment:
are we filling up the instanxce ids here?

Reply:
we should go on retrying till we get the number of instance ids equal to the number of replicas

Reply:
I think this is taken care of in the `InstancePoolLifecycleStateProvisioning` section. Based on this https://docs.oracle.com/en-us/iaas/Content/Compute/Concepts/instancemanagement.htm#Instance__lifecycle-states-instance-pools I read it to mean the state will only become running once all the instances are created and running. So it won't reach that state until that happens.

Reply:
> are we filling up the instanxce ids here?

After talking with Shyam we will create a second PR to handle updating the `ProviderIDList` after we land this one.

- [ ] Cut new issue to address this

Reply:
#90 has been created.